src/python/xml_parser-py.c

```diff
@@ -792,3 +790,247 @@ py_xml_parse_main_metadata_together(G_GNUC_UNUSED PyObject *self, PyObject *args
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    // TODO: correct thing to do with error?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    Py_XINCREF(py_newpkgcb);
+    Py_XINCREF(py_warningcb);
+    Py_XINCREF(cbdata.py_pkgs);
+
+    cr_XmlParserNewPkgCb ptr_c_newpkgcb = NULL;
+    cr_XmlParserWarningCb ptr_c_warningcb = NULL;
+
+    if (py_newpkgcb != Py_None)
+        ptr_c_newpkgcb = c_newpkgcb;
+    if (py_warningcb != Py_None)
+        ptr_c_warningcb = c_warningcb;
+
+    cbdata.py_newpkgcb = py_newpkgcb;
+    cbdata.py_pkgcb = NULL;  // TODO: What should go here?
+    cbdata.py_warningcb = py_warningcb;
+    cbdata.py_pkgs = PyDict_New();
+
+    self->pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, ptr_c_newpkgcb, &cbdata, ptr_c_warningcb, &cbdata, &tmp_err);
+
+    // TODO: correct?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    if (self->pkg_iterator == NULL) {
+        PyErr_SetString(CrErr_Exception, "PkgIterator initialization failed");
+        return -1;
+    }
+
+    Py_XDECREF(py_newpkgcb);
+    Py_XDECREF(py_warningcb);
+    Py_XDECREF(cbdata.py_pkgs);
+
+    return 0;
+}
+
+static void
+pkg_iterator_dealloc(_PkgIteratorObject *self)
+{
+    GError *tmp_err;
+    if (self->pkg_iterator)
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    Py_TYPE(self)->tp_free(self);
+    // TODO: do something with error?
+}
+
+// static PyObject *
+// pkg_iterator_repr(_PkgIteratorObject *self)
+// {
+//     if (check_PkgIteratorStatus(self))
+//         return NULL;
+//     PyObject *repr;
+//     if (self->pkg_iterator) {
+//         repr = PyUnicode_FromFormat("<createrepo_c.PkgIterator object>");
+//     } else {
+//         repr = PyUnicode_FromFormat("<createrepo_c.PkgIterator object id -, ->");
+//     }
+//     return repr;
+// }
+
+// static PyObject *
+// pkg_iterator_str(_PackageObject *self)
+// {
+//     PyObject *ret;
+//     if (check_PkgIteratorStatus(self))
+//         return NULL;
+//     if (self->package) {
+//         char *nevra = cr_package_nvra(self->package);
+//         ret = PyUnicode_FromString(nevra);
+//         free(nevra);
+//     } else {
+//         ret = PyUnicode_FromString("-");
+//     }
+//     return ret;
+// }
+
+static PyObject *
+pkg_iterator_next_package(_PkgIteratorObject *self, G_GNUC_UNUSED void *nothing)
+{
+    cr_Package *pkg;
+    GError *tmp_err = NULL;
+
+    if (check_PkgIteratorStatus(self)) {
+        return NULL;
+    }
+    pkg = cr_PkgIterator_parse_next(self->pkg_iterator, &tmp_err);
+    if (tmp_err) {
+        cr_package_free(pkg);
+        // TODO: correct?
+        nice_exception(&tmp_err, NULL);
+        return NULL;
+    }
+
+    if (!pkg) {
+        assert(cr_PkgIterator_is_finished(self->pkg_iterator));
+        PyErr_SetNone(PyExc_StopIteration);
+        return NULL;
+    }
+
+    return Object_FromPackage(pkg, 1);
```

Comment:
Finally given that we don't use `c_pkgcb` we should do a similar dance as it does with the `py_pkgs` dictionary.
So that if the user passes in a package via `newpkgcb` he gets back the same one (filled with data).
```suggestion
    PyObject *keyFromPtr = PyLong_FromVoidPtr(pkg);
    PyObject *py_pkg = PyDict_GetItem(self->cbdata->py_pkgs, keyFromPtr);
    if (py_pkg) {
        // Remove pkg from PyDict but keep one reference so its not freed if the
        // user doesn't have any references to the package
        Py_XINCREF(py_pkg);
        PyDict_DelItem(self->cbdata->py_pkgs, keyFromPtr);
    } else {
        // The package was not provided by user in c_newpkgcb,
        // create new python package object
        py_pkg = Object_FromPackage(pkg, 1);
    }
    Py_DECREF(keyFromPtr);
    return py_pkg;
```